ion-hash/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Implements the [Ion Hash specification](https://amzn.github.io/ion-hash/docs/spec.html)
+//!
+//! ## Examples
+//! ```rust
+//! use ion_rs::value::loader::{self, Loader};
+//! use ion_rs::result::IonResult;
+//!
+//! # fn main() -> IonResult<()> {
+//!   let loader = loader::loader();
+//!   let elem = loader.iterate_over(b"\"hello world\"")?.next().unwrap()?;
+//!   let digest = ion_hash::sha256(&elem);
+//!   println!("{:?}", digest);
+//! # Ok(())
+//! # }
+//! ```
+
+use ion_rs::{value::Element, IonType};
+
+use sha2::{digest::Output, Digest, Sha256};
+
+/// Utility to hash an [`Element`] using SHA-256 as the hash function.
+pub fn sha256<E: Element>(elem: &E) -> Vec<u8> {
+    let mut hasher = IonHasher::new(Sha256::new());
+    let result = hasher.hash_element(elem);
+    Vec::from(result.as_slice())
+}
+
+/// Bytes markers as per the spec.
+struct Markers;
+impl Markers {
+    /// single byte begin marker
+    const B: u8 = 0x0B;
+    /// single byte end marker
+    const E: u8 = 0x0E;
+    /// single byte escape
+    const ESC: u8 = 0x0C;
+    /// type qualifier octet consisting of a four-bit type code T followed by a four-bit qualifier Q
+    /// (this varies per type)
+    const TQ_STRING: u8 = 0x80;
+}
+
+// TODO: In the fullness of time, we will have a IonHashReader and a
+// IonHashWriter. This will allow reading/writing a value *while* computing a
+// hash. For now, we provide a standalone hasher using the Element API.
+/// Provides Ion hash over arbitrary [`Element`] instances with a given
+/// [`Digest`] algorithm.
+pub struct IonHasher<D>
+where
+    D: Digest,
+{
+    hasher: D,
+}
+
+impl<D> IonHasher<D>
+where
+    D: Digest,
+{
+    fn new(hasher: D) -> Self {
+        Self { hasher }
+    }
+
+    /// Computes the Ion hash over an [`Element`] recursively using this
+    /// hasher's [`Digest`]
+    fn hash_element<E: Element + ?Sized>(&mut self, elem: &E) -> Output<D> {
+        let serialized_bytes = match elem.ion_type() {
+            IonType::Null => todo!(),
+            IonType::Boolean => todo!(),
+            IonType::Integer => todo!(),
+            IonType::Float => todo!(),
+            IonType::Decimal => todo!(),
+            IonType::Timestamp => todo!(),
+            IonType::Symbol => todo!(),
+            IonType::String => self.visit_string(elem.as_str().unwrap()),
```

Comment:
I missed this earlier, but shouldn't this API return `Result<T, E>` or some such to avoid the unwrap?  At a bare minimum, a TODO screaming at us to fix this would be good because it is a bad idea for a library to panic this way.

Reply:
Fixed in 00d23d4